include/zephyr/net/ieee802154_ie.h

```diff
@@ -0,0 +1,338 @@
+ * parts of the types required for IE support that need to be visible to IEEE
+ * 802.15.4 drivers and L2 at the same time, i.e. everything related to header
+ * IE representation, parsing and generation. Types visible to L2 only can be
+ * found in the subsystem folder in ieee802154_frame_ie.h
+ *
+ * All specification references in this file refer to IEEE 802.15.4-2020.
+ *
+ * @note All structs and attributes in this file that directly represent parts
+ * of IEEE 802.15.4 frames are in LITTLE ENDIAN, see section 4, especially
+ * section 4.3.
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_IEEE802154_IE_H_
+#define ZEPHYR_INCLUDE_NET_IEEE802154_IE_H_
+
+#include <zephyr/net/buf.h>
+#include <zephyr/sys/byteorder.h>
+
+/**
+ * @addtogroup ieee802154_driver
+ * @{
+ *
+ * @name IEEE 802.15.4, section 7.4.2: MAC header information elements
+ * @{
+ */
+
+/**
+ * @brief Information Element Types.
+ *
+ * @details See sections 7.4.2.1 and 7.4.3.1.
+ */
+enum ieee802154_ie_type {
+	IEEE802154_IE_TYPE_HEADER = 0x0,
+	IEEE802154_IE_TYPE_PAYLOAD,
+};
+
+/**
+ * @brief Header Information Element IDs.
+ *
+ * @details See section 7.4.2.1, table 7-7, partial list, only IEs actually used
+ * are implemented.
+ */
+enum ieee802154_header_ie_element_id {
+	IEEE802154_HEADER_IE_ELEMENT_ID_VENDOR_SPECIFIC_IE = 0x00,
+	IEEE802154_HEADER_IE_ELEMENT_ID_CSL_IE = 0x1a,
+	IEEE802154_HEADER_IE_ELEMENT_ID_RIT_IE = 0x1b,
+	IEEE802154_HEADER_IE_ELEMENT_ID_RENDEZVOUS_TIME_IE = 0x1d,
+	IEEE802154_HEADER_IE_ELEMENT_ID_TIME_CORRECTION_IE = 0x1e,
+	IEEE802154_HEADER_IE_ELEMENT_ID_HEADER_TERMINATION_1 = 0x7e,
+	IEEE802154_HEADER_IE_ELEMENT_ID_HEADER_TERMINATION_2 = 0x7f,
+	/* partial list, add additional ids as needed */
+};
+
+/** @cond INTERNAL_HIDDEN */
+#define IEEE802154_VENDOR_SPECIFIC_IE_OUI_LEN 3
+/** INTERNAL_HIDDEN @endcond */
+
+/** @brief Vendor Specific Header IE, see section 7.4.2.3. */
+struct ieee802154_header_ie_vendor_specific {
+	uint8_t vendor_oui[IEEE802154_VENDOR_SPECIFIC_IE_OUI_LEN];
+	uint8_t *vendor_specific_info;
+} __packed;
+
+/** @brief Full CSL IE, see section 7.4.2.3. */
+struct ieee802154_header_ie_csl_full {
+	uint16_t csl_phase;
+	uint16_t csl_period;
+	uint16_t csl_rendezvous_time;
+} __packed;
+
+/** @brief Reduced CSL IE, see section 7.4.2.3. */
+struct ieee802154_header_ie_csl_reduced {
+	uint16_t csl_phase;
+	uint16_t csl_period;
+} __packed;
+
+/** @brief Generic CSL IE, see section 7.4.2.3. */
+struct ieee802154_header_ie_csl {
+	union {
+		struct ieee802154_header_ie_csl_reduced reduced;
+		struct ieee802154_header_ie_csl_full full;
+	};
+} __packed;
+
+/** @brief RIT IE, see section 7.4.2.4. */
+struct ieee802154_header_ie_rit {
+	uint8_t time_to_first_listen;
+	uint8_t number_of_repeat_listen;
+	uint16_t repeat_listen_interval;
+} __packed;
+
+/** @brief Rendezvous Time IE, see section 7.4.2.6. */
+struct ieee802154_header_ie_rendezvous_time {
+	uint16_t rendezvous_time;
+	uint16_t wakeup_interval;
```

Comment:
Correct. Fixed. The standard has a bug there because it only indicates the field being optional in the text, not in figure 7-30 (0/2 as field length) as it should.

In case you wonder: I only introduce instantiation helpers (defines) for those IEs that actually need to be instantiated already. I'll add more if we need to instantiate (as compared to just "parse") additional IEs. Hope that's ok?